example-client/example-client.js

```diff
@@ -26,27 +26,87 @@ const FAKE_STATE = crypto.randomBytes(64).toString('hex');
 
 const app = express();
 
-// Exchanges an authorization code for an authorization token
-function exchangeCodeForAuthToken(code, callback) {
+// A real implementation would use session middleware; because we're
+// an example app, we'll just use a global object.
+let session = {};
+
+// This posts to the token endpoint, either to exchange an authorization
+// code for an access token, or to obtain a new access token from a
+// refresh token (depending on the contents of the given payload).
+//
+// It then sets the session data accordingly: if the token request failed,
+// the user (if any) is logged out. Otherwise, the user is logged in.
+function postToTokenUrl(payload, callback) {
   request.post(UAA_TOKEN_URL, {
-    form: {
-      code,
-      grant_type: 'authorization_code',
-      response_type: 'token',
+    form: payload,
+  }, function(err, response, body) {
+    if (!err && response && response.statusCode !== 200) {
+      err = new Error(`got HTTP ${response.statusCode}`);
+    }
+    if (err) {
+      session = {};
+    } else {
+      const responseBody = JSON.parse(body);
+      const decodedToken = jwt.decode(responseBody.access_token);
+
+      Object.assign(session, {
+        email: decodedToken.email,
+        refreshToken: responseBody.refresh_token,
+        expiry: Date.now() + responseBody.expires_in * 1000
+      });
+    }
+    callback(err);
+  });
+}
+
+// This middleware detects if the user's access token has expired; if
+// so, it attempts to renew it with the server. This is primarily a
+// security measure, to ensure that users who lose access to the system
+// are logged out of it ASAP.
+app.use(function tokenRefreshMiddleware(req, res, next) {
+  if (session.expiry && Date.now() > session.expiry) {
+    // Our session has expired! Let's renew it.
+    console.log('User session has expired, attempting to renew it.');
+    postToTokenUrl({
+      grant_type: 'refresh_token',
+      refresh_token: session.refreshToken,
       client_id: CLIENT_ID,
       client_secret: CLIENT_SECRET,
-    },
-  }, callback);
-}
+    }, (err) => {
+      if (err) {
+        console.log('Renewal unsuccessful, user was logged out.');
+      } else {
+        console.log('Renewal successful.');
+      }
+      next();
+    });
+  } else {
+    next();
+  }
+});
 
-// Our "home page" contains a login link.
+// Our "home page".
 app.get('/', (req, res) => {
-  const url = UAA_AUTH_URL + '?' + querystring.stringify({
-    'client_id': CLIENT_ID,
-    'state': FAKE_STATE,
-    'response_type': 'code'
-  });
-  res.send(`<a href="${url}">Log in</a> via <code>${UAA_AUTH_URL}</code>!`);
+  function getLoggedInHtml() {
```

Comment:
Very minor: I found it a little hard to parse what exactly the route handler was doing and think it would be clearer if just the `res.send(session.email ? getLoggedInHtml() : getLoggedOutHtml());` was in the handler and the two functions were moved outside.